### DISCUSSION TITLE
OnAfterRender getting called multiple times

### DISCUSSION THREAD

**Discussion by SSzretter** [oqtane/oqtane.framework#4801] [Q&A]
My OnAfterRender is getting called multiple times and this is causing issues with several javascript libraries working properly including BoldReports, Syncfusion and FormIO.   In the case of BoldReports and Syncfusion I opened tickets with them and provided sample module projects and heard back in one case that they were not exactly sure what is wrong but that it seemed like the components were getting created twice.    

I found that at least on possible issue is related to the on after render being called multiple times.  Tracing, it hit my trace 9 times.   I have some code that looks for first render = false and some other conditions to reduce the number of calls but I still get at least 2 calls so then I had to try to put code in the javascript side to detect if my javascript object was already created for Bold/ForimIO etc.  but that does not seem to be reliable for some reason. 

I experimented with the settings in the admin dashboard in oqtane and tried setting it to interactive instead of static and to not prerender because it seems blazor pre-renders and calls on after render during that phase and then later when it renders the server interactive side.   I confirmed by tracing that the mode is interactive at least.   No difference.

I am including my resources within the specific razor page using new Resouce - I have tried including them in moduleinfo as well and that did not seem to matter.   It seems like all the scripts and css are loading though.

Here are a couple sample modules, as stripped down as I could make them.   I appreciate any suggestions you might have- especially if I am loading the resources properly, handling the on after render / js interop properly.   I am having issues in a FormIO solution I am testing as well and it feels similar so I think if I can get it fixed in one I can fix it across the board.

https://github.com/SSzretter/SPSITECH.Module.Oqtane.BoldReports
https://github.com/SSzretter/SPSITECH.Module.SyncfusionNotifications

**Comment by SSzretter** [comment]
By the way this is what BoldReports said about the sample project when they dug in:

![image](https://github.com/user-attachments/assets/ee5eeb0d-7a91-4e74-a167-60a119f147bf)

![image](https://github.com/user-attachments/assets/132f77d9-44f2-4c54-b283-6840fdbc97a4)

**Reply by SSzretter** [reply]
On the SyncFusion side I was focusing on toasts specifically and they recommended adding an on timer delay in javascript.   I find that that helps but is not 100% reliable.   I have seen similar issues to whats happening in the BoldReports example though - the Syncfusion Grid is a good one, it does not always load on the first page draw and you have to refresh, I am having this issue with FormIO as well where it does not draw on first or sometimes multiple refreshes and other issues.

**Comment by sbwalker** [comment]
Integrating JavaScript with Blazor can be very challenging.
